File: src/solutions/2024/09/ts/main.ts
import { argv } from "node:process";
import type { SolutionOptions } from "src/options.ts";
import { run } from "src/run.ts";

const options: SolutionOptions = {
   hasAlternate: false,
   hasIo: false,
};

type Disk = {
   size: number;
   free: number;
   moved: number;
   space: number[];
};

function solve(input: string, p2: boolean): string {
   const disks: Disk[] = [];
   for (let i = 0; i < input.length; i += 2) {
      if (i + 1 === input.length) {
         disks.push({ size: +input[i], free: 0, space: [], moved: 0 });
      }
      else {
         disks.push({
            size: +input[i],
            free: +input[i + 1],
            space: [],
            moved: 0,
         });
      }
   }

   let left = 0;
   let right = disks.length - 1;
   let firstPos = 0;
   if (p2) {
      while (left < right) {
         if (disks[left].free >= disks[right].size) {
            for (let i = 0; i < disks[right].size; i++) {
               disks[left].space.push(right);
            }
            disks[left].free -= disks[right].size;
            disks[right].moved += disks[right].size;
            disks[right].size = 0;
            left = firstPos;
            right--;
         }
         else {
            left++;
         }

         if (disks[firstPos].free === 0) {
            firstPos = left;
         }

         if (left === right) {
            left = firstPos;
            right--;
         }
      }
   }
   else {
      while (left < right) {
         if (disks[left].free > 0 && disks[right].size > 0) {
            disks[left].space.push(right);
            disks[left].free--;
            disks[right].size--;
         }

         if (disks[right].size === 0) {
            right--;
         }
         if (disks[left].free === 0) {
            left++;
         }
      }
   }

   let total = 0;
   let i = 0;
   for (const k in disks) {
      i += disks[k].moved;
      for (let s = 0; s < disks[k].size; s++) {
         total += i * +k;
         i++;
      }
      for (const n of disks[k].space) {
         total += i * +n;
         i++;
      }
      i += disks[k].free;
   }

   return total.toString();
}

function part1(input: string, _isTest: boolean): string {
   return solve(input, false);
}

function part2(input: string, _isTest: boolean): string {
   return solve(input, true);
}

if (import.meta.main) {
   run(argv, part1, part2, options);
}

File: src/solutions/2024/09/csharp/Main.cs

using System.Drawing;
using Core;

namespace Year2024;

public static class Day09
{
   static SolutionOptions OPTIONS = new()
   {
      HasAlternate = false,
      HasIO = false,
   };

   private struct Disk
   {
      public long size;
      public long free;
      public long moved;
      public long[] space;
   }

   static private string Solve(string input, bool p2)
   {
      Disk[] disks = input
         .ToCharArray()
         .Where((_, i) => i % 2 == 0)
         .Select((c, i) => i * 2 + 1 == input.Length
            ? new Disk
            {
               size = (long)char.GetNumericValue(c),
               free = 0,
               space = [],
               moved = 0
            }
            : new Disk
            {
               size = (long)char.GetNumericValue(c),
               free = (long)char.GetNumericValue(input[i * 2 + 1]),
               space = new long[(long)char.GetNumericValue(input[i * 2 + 1])],
               moved = 0
            })
         .ToArray();

      var left = 0;
      var right = disks.Length - 1;
      var firstPos = 0;
      if (p2)
      {
         while (left < right)
         {
            if (disks[left].free >= disks[right].size)
            {
               for (var j = 0; j < disks[right].size; j++)
               {
                  disks[left].space[disks[left].space.Length - disks[left].free] = right;
                  disks[left].free--;
                  disks[right].moved++;
               }
               disks[right].size = 0;
               left = firstPos;
               right--;
            }
            else
            {
               left++;
            }

            if (disks[firstPos].free == 0)
            {
               firstPos = left;
            }

            if (left == right)
            {
               left = firstPos;
               right--;
            }
         }
      }
      else
      {
         while (left < right)
         {
            if (disks[left].free > 0 && disks[right].size > 0)
            {
               disks[left].space[disks[left].space.Length - disks[left].free] = right;
               disks[left].free--;
               disks[right].size--;
            }

            if (disks[right].size == 0)
            {
               right--;
            }
            if (disks[left].free == 0)
            {
               left++;
            }
         }
      }

      long total = 0;
      long i = 0;
      for (long k = 0; k < disks.Length; k++)
      {
         i += disks[k].moved;
         for (var s = 0; s < disks[k].size; s++)
         {
            total += i * k;
            i++;
         }
         for (var s = 0; s < disks[k].space.Length - disks[k].free; s++)
         {
            total += i * disks[k].space[s];
            i++;
         }
         i += disks[k].free;
      }

      return total.ToString();
   }

   static string Part1(string input, bool isTest = false)
   {
      return Solve(input, false);
   }

   static string Part2(string input, bool isTest = false)
   {
      return Solve(input, true);
   }

   static void Main(string[] args)
   {
      Runner.Run(args, Part1, Part2, OPTIONS);
   }
}
